### PR TITLE
Check known recipes

### LINKED ISSUE #13: Use my API key to check if I know the recipe
This one is probably an "I'd love a pull request" issue; it also opens potential for discussion of cross-platform configuration files, which is usually not a favorite topic. Opening this so it's available for discussion.

I might work on this when I'm more comfortable with Rust - I do not know enough yet to get file caching for #7 working successfully.

### PR DESCRIPTION
**Pull request by Iiridayn**
Fixes #13.

Not 100% yet; only shows the recipe id, not the id + name of the item to unlock the recipe. Still, good enough to be useful. Be aware there is some kind of cache on the API side, so an unlock may not immediately show up. Also doesn't seem possible to check discovery recipes (though, free). Also also, doesn't check if you have a character of the appropriate discipline + crafting level for non-unlocked recipes.

All that aside, already quite useful.

Based on the min sale branch as otherwise merging it would produce a conflict, which doesn't seem worth it. Any concerns re that branch?

### CODE REVIEW

**Comment by Iiridayn** [comment]
Should probably also gracefully handle wrong API keys and keys w/o permissions eventually.

**Comment by Iiridayn** [comment]
There we go. Ready for review - especially let me know what you think of the config file location, etc, as we can put some of the switches in there later.

File: src/main.rs [hunk 1]
```diff
@@ -462,7 +529,13 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
                 .join("/"),
             item_id,
-            recipe_id: recipe.id,
+            unknown_recipes: profitable_item.used_recipes.iter().filter_map(|(id, known)| {
+                if !*known {
+                    Some(*id)
+                } else {
+                    None
+                }
+            }).collect(),
```
Comment (t-mw) [inline src/main.rs:538]: It can be convenient to dereference in the arguments:
```suggestion
            unknown_recipes: profitable_item.used_recipes.iter().filter_map(|(&id, &known)| {
                if !known {
                    Some(id)
                } else {
                    None
                }
            }).collect(),
```

File: src/main.rs [hunk 2]
```diff
@@ -296,6 +340,28 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+        let unknown_recipes: Vec<u32> = profitable_item.used_recipes.iter().filter_map(|(id, known)| {
+            if !*known {
+                Some(*id)
+            } else {
+                None
+            }
+        }).collect();
+        if unknown_recipes.len() > 0 {
+            println!(
+                "You {} craft this yet. Missing recipe id{}: {}",
```
Comment (t-mw) [inline src/main.rs:352]: This might make more sense when an api key isn't being used:
```suggestion
                "You {} craft this yet. Requires recipe id{}: {}",
```

File: src/crafting.rs [hunk 3]
```diff
@@ -495,23 +528,40 @@ impl Listing {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+enum RecipeSource {
+    Automatic,
+    Discovered,
+    Purchase,
```
Comment (t-mw) [inline src/crafting.rs:535]: `Discovered` and `Purchased` are a bit confusing to me when I read them out of context, because it sounds like it refers to the recipe being known. How about `Discoverable` and `Purchasable`?

File: src/crafting.rs [hunk 4]
```diff
@@ -231,13 +237,37 @@ pub fn calculate_crafting_profit(
...
+                    }
+                    _ => {
+                        if let Some(known_recipes) = known_recipes {
+                            used_recipes.insert(id, known_recipes.contains(&id));
+                        } else {
+                            // If we have no known recipes, assume we know none
+                            used_recipes.insert(id, false);
+                        }
+                    }
+                }
```
Comment (t-mw) [inline src/crafting.rs:266]: Do you mean that the recipes are not included in the api, because they can't be purchased? If that's the case would it make sense to just check `Purchased`?

```suggestion
                match recipe.source {
                    RecipeSource::Purchased => {
                        if let Some(known_recipes) = known_recipes {
                            used_recipes.insert(id, known_recipes.contains(&id));
                        } else {
                            // If we have no known recipes, assume we know none
                            used_recipes.insert(id, false);
                        }
                    }
                    // Not included in the API
                    _ => {
                        // TODO: check if account has a char with the required crafting level
                        used_recipes.insert(id, true);
                    }
                }
```

File: src/crafting.rs [hunk 5]
```diff
@@ -148,6 +150,9 @@ fn calculate_precise_min_crafting_cost(
                     return None;
                 }
             }
+
+            used_recipe(&recipe);
```
Comment (t-mw) [inline src/crafting.rs:154]: I think that this marks a recipe as used even when the item to craft has been purchased from the trading post. If that works for you, then maybe just something to note in a comment.

Only adding recipes for items that need crafting would be a bit more complicated, but I think it could be done in a similar way to how `context.purchases` is generated, but just with inverted conditions. Something like:

```
@@ -98,6 +98,7 @@ struct PreciseCraftingCost {
 
 struct PreciseCraftingCostContext {
     purchases: Vec<(u32, Rational32, Source)>,
+    crafted: Vec<u32>,
     crafting_steps: Rational32,
 }
 
@@ -119,6 +120,7 @@ fn calculate_precise_min_crafting_cost(
         Rational32::from(recipe.map(|recipe| recipe.output_item_count).unwrap_or(1));
 
     let purchases_ptr = context.purchases.len();
+    let crafted_ptr = context.crafted.len();
     let crafting_steps_before = context.crafting_steps;
 
     let crafting_cost = recipe.and_then(|recipe| {
@@ -181,6 +183,11 @@ fn calculate_precise_min_crafting_cost(
 
     if source == Source::Crafting {
         context.crafting_steps += item_count / output_item_count;
+        for crafted_item_id in context.crafted.drain(crafted_ptr..) {
+            if let Some(recipe) = recipes_map.get(&crafted_item_id) {
+                used_recipe(recipe);
+            }
+        }
     } else {
         for (purchase_id, purchase_quantity, purchase_source) in
             context.purchases.drain(purchases_ptr..)
@@ -197,6 +204,8 @@ fn calculate_precise_min_crafting_cost(
 
     if source != Source::Crafting {
         context.purchases.push((item_id, item_count, source));
+    } else {
+        context.crafted.push(item_id);
     }
     if source == Source::TradingPost {
         tp_listings_map
```

File: src/crafting.rs [hunk 6]
```diff
@@ -231,13 +237,37 @@ pub fn calculate_crafting_profit(
     let mut total_crafting_cost = Rational32::zero();
     let mut crafting_count = 0;
     let mut total_crafting_steps = Rational32::zero();
+    let mut used_recipes = HashMap::new();
```
Comment (t-mw) [inline src/crafting.rs:240]: Could `used_recipes` be a `HashSet`?

**Review by t-mw — COMMENTED** [review]
Generally looks good, thank you! Just a few comments.

File: src/crafting.rs (hunk 5 above)
Comment (t-mw) [reply src/crafting.rs:154]:
```diff
@@ -98,6 +98,7 @@ struct PreciseCraftingCost {
 
 struct PreciseCraftingCostContext {
     purchases: Vec<(u32, Rational32, Source)>,
+    crafted: Vec<u32>,
     crafting_steps: Rational32,
 }
 
@@ -108,7 +109,6 @@ fn calculate_precise_min_crafting_cost(
     item_count: Rational32,
     recipes_map: &HashMap<u32, Recipe>,
     items_map: &HashMap<u32, api::Item>,
-    used_recipe: &mut dyn FnMut(&Recipe) -> (),
     tp_listings_map: &mut BTreeMap<u32, ItemListings>,
     context: &mut PreciseCraftingCostContext,
     opt: &CraftingOptions,
@@ -119,6 +119,7 @@ fn calculate_precise_min_crafting_cost(
         Rational32::from(recipe.map(|recipe| recipe.output_item_count).unwrap_or(1));
 
     let purchases_ptr = context.purchases.len();
+    let crafted_ptr = context.crafted.len();
     let crafting_steps_before = context.crafting_steps;
 
     let crafting_cost = recipe.and_then(|recipe| {
@@ -135,7 +136,6 @@ fn calculate_precise_min_crafting_cost(
                     ingredient_count,
                     recipes_map,
                     items_map,
-                    used_recipe,
                     tp_listings_map,
                     context,
                     opt,
@@ -151,8 +151,6 @@ fn calculate_precise_min_crafting_cost(
                 }
             }
 
-            used_recipe(&recipe);
-
             Some(cost)
         }
     });
@@ -180,6 +178,7 @@ fn calculate_precise_min_crafting_cost(
     };
 
     if source == Source::Crafting {
+        context.crafted.push(item_id);
         context.crafting_steps += item_count / output_item_count;
     } else {
         for (purchase_id, purchase_quantity, purchase_source) in
@@ -192,6 +191,7 @@ fn calculate_precise_min_crafting_cost(
                     .pending_buy_quantity -= purchase_quantity;
             }
         }
+        context.crafted.drain(crafted_ptr..);
         context.crafting_steps = crafting_steps_before;
     }
 
@@ -278,6 +281,7 @@ pub fn calculate_crafting_profit(
 
         let mut context = PreciseCraftingCostContext {
             purchases: vec![],
+            crafted: vec![],
             crafting_steps: Rational32::zero(),
         };
 
@@ -289,7 +293,6 @@ pub fn calculate_crafting_profit(
             output_item_count.into(),
             recipes_map,
             items_map,
-            &mut used_recipe,
             &mut tp_listings_map,
             &mut context,
             opt,
@@ -321,6 +324,12 @@ pub fn calculate_crafting_profit(
             break;
         }
 
+        for item_id in &context.crafted {
+            if let Some(recipe) = recipes_map.get(item_id) {
+                used_recipe(recipe);
+            }
+        }
+
         for (purchase_id, count, purchase_source) in &context.purchases {
             if *purchase_source != Source::TradingPost {
                 continue;
```

File: src/main.rs [hunk 7]
```diff
@@ -417,7 +484,7 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         "Name",
         "Disciplines",
         "Item id",
-        "Recipe id",
+        "Unk Recipe id",
```
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/main.rs:487]: It's blank if they have the recipe unlocked though. Maybe "recipes to learn"?

I fixed it so those w/o an API key it still assumes they can craft anything with an automatic unlock - though some common cheap recipes are still assumed to be unknown.

File: src/crafting.rs (hunk 6 above)
Comment (Iiridayn) [reply src/crafting.rs:240]: I chose a HashMap because I wanted to track used recipes which are known and unknown. You are right though that I ultimately only care about unknown recipes - and I know which are which when I insert them. Good change.

File: src/main.rs (hunk 1 above)
Comment (Iiridayn) [reply src/main.rs:538]: Good to learn. Resolved by using a HashSet.

File: src/request.rs [hunk 8]
```diff
@@ -158,3 +160,14 @@ where
...
 }
+
+pub async fn fetch_account_recipes(key: &str) -> Result<HashSet<u32>, Box<dyn std::error::Error>> {
+    let url = format!("https://api.guildwars2.com/v2/account/recipes?access_token={}", key);
+    println!("Fetching {}", url);
+    let result = reqwest::get(url).await?;
+    if result.status() != 200 {
+        let err: serde_json::value::Value = result.json().await?;
+        panic!("API error fetching recipe unlocks: {}", err["text"]);
+    }
```
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/request.rs:171]: Much nicer, thank you.

File: src/request.rs [hunk 9]
```diff
@@ -158,3 +160,14 @@ where
 
     Ok(result)
 }
+
+pub async fn fetch_account_recipes(key: &str) -> Result<HashSet<u32>, Box<dyn std::error::Error>> {
+    let url = format!("https://api.guildwars2.com/v2/account/recipes?access_token={}", key);
+    println!("Fetching {}", url);
```
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/request.rs:166]: I don't feel like it's a noteworthy privacy risk and could aid users to fix errors, but we may as well not print it. The error handling code you supplied is tremendously better than what I had in place anyway.

File: src/crafting.rs [hunk 10]
```diff
@@ -495,23 +528,40 @@ impl Listing {
...
 
 impl From<api::Recipe> for Recipe {
     fn from(recipe: api::Recipe) -> Self {
+        let source = if recipe.is_purchased() {
+            RecipeSource::Purchase
+        } else if recipe.is_automatic() {
+            RecipeSource::Automatic
+        } else {
+            RecipeSource::Discovered
+        };
```
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/crafting.rs:557]: Part of the problem is the crafting::Recipe inherits from two different recipe sources - `api::Recipe` and `gw2efficiency::Recipe`. You'll note just a few lines down from there the `Achievement` recipe source is used coming from `gw2efficiency::Recipe`.

[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/crafting.rs:557]: Unless you're saying to put it on `api::Recipe`? The enum properly belongs here, due to the aforementioned merging of two recipe sources. I pulled out `is_purchased` and `is_automatic` to keep what implementation details I could in that file.

File: src/crafting.rs [hunk 11]
```diff
@@ -231,13 +237,37 @@ pub fn calculate_crafting_profit(
...
     let mut min_sell = 0;
     let max_sell = tp_listings_map.get(&item_id)
         .unwrap_or_else(|| panic!("Missing listings for item id: {}", item_id))
         .buys.last().map_or(0, |l| l.unit_price);
     let mut breakeven = Rational32::zero();
 
+    // Check if we can craft this
+    let mut used_recipe = |recipe: &Recipe| {
+        if let Some(id) = recipe.id {
+            if !&used_recipes.contains_key(&id) {
```
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/crafting.rs:251]: Yeah, too much there, and readable.

File: src/crafting.rs (hunk 4 above)
Comment (Iiridayn) [reply src/crafting.rs:266]: Some are awarded by an achievement instead. I've make it more explicit by specifying all branches of the enum.

File: src/crafting.rs (hunk 5 above)
Comment (Iiridayn) [reply src/crafting.rs:154]: Yeah, I moved it down. I originally didn't want to as I figured I'd need to pull recipe out of the scope, and overlooked it or simply noted that every example I'd seen was account bound anyway. Looking through the logic, `unwrap` should be fine if we get to that point.

File: src/main.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Iiridayn) [reply src/main.rs:485]: Came up w/the same idea trying to fix it. Pushed.

File: src/crafting.rs (hunk 5 above)
Comment (t-mw) [reply src/crafting.rs]: It's a bit hard to follow in my code, but https://github.com/t-mw/gw2-arbitrage/blob/master/src/crafting.rs#L178 doesn't necessarily mean that the item will be crafted if it's an ingredient, because it might be rolled back by a parent recipe in https://github.com/t-mw/gw2-arbitrage/blob/master/src/crafting.rs#L190. I think the changes in my last comment should work for that case, I'll try adding it after you merge.

**Review by t-mw — APPROVED** [review]
Looks good! Thanks again
